### PR TITLE
Implement key backup APIs for rust and create backup in `bootstrapSecretStorage`

### PR DESCRIPTION
**Pull request by BillCarsonFr**
Fixes https://github.com/vector-im/crypto-internal/issues/107
~~Based on https://github.com/matrix-org/matrix-js-sdk/pull/3689~~

Implements missing key backup API in rust and create backups as part of the bootstrap security process
## Checklist

-   [x] Tests written for new code (and old code if feasible)
-   [x] Linter and other CI checks pass
-   [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-js-sdk/blob/develop/CONTRIBUTING.md))

---
This change is marked as an *internal change* (Task), so will not be included in the changelog.

### CODE REVIEW

File: spec/integ/crypto/crypto.spec.ts [hunk 1]
```diff
@@ -55,6 +54,7 @@ import {
     Room,
     RoomMember,
     RoomStateEvent,
+    CryptoEvent,
```
Comment (richvdh) [inline spec/integ/crypto/crypto.spec.ts:57]: this was sorted before. :-1:  to this change.

File: spec/integ/crypto/crypto.spec.ts [hunk 2]
```diff
@@ -2202,9 +2202,11 @@ describe.each(Object.entries(CRYPTO_BACKENDS))("crypto (%s)", (backend: string,
                     "express:/_matrix/client/v3/user/:userId/account_data/:type(m.secret_storage.*)",
                     (url: string, options: RequestInit) => {
                         const content = JSON.parse(options.body as string);
+
                         if (content.key) {
                             resolve(content.key);
                         }
+
```
Comment (richvdh) [inline spec/integ/crypto/crypto.spec.ts:2209]: not convinced this is an improvement, but ymmv

File: spec/integ/crypto/crypto.spec.ts [hunk 3]
```diff
@@ -2457,17 +2455,25 @@ describe.each(Object.entries(CRYPTO_BACKENDS))("crypto (%s)", (backend: string,
...
+            expect(signatures).toBeDefined();
+            expect(signatures![aliceClient.getUserId()!]).toBeDefined();
+            const mskId = await aliceClient.getCrypto()!.getCrossSigningKeyId(CrossSigningKey.Master)!;
+            expect(signatures![aliceClient.getUserId()!][`ed25519:${mskId}`]).toBeDefined();
         });
 
-        oldBackendOnly("Reset key backup should create a new backup and update 4S", async () => {
-            // First set up 4S and key backup
+        it("Reset key backup should create a new backup and update 4S", async () => {
+            // First set up recovery
```
Comment (richvdh) [inline spec/integ/crypto/crypto.spec.ts:2476]: :-1:

**Review by richvdh — CHANGES_REQUESTED** [review]

File: src/rust-crypto/rust-crypto.ts [hunk 4]
```diff
@@ -13,7 +13,11 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */
+/**
+ * Utilities common to olm encryption algorithms
+ */
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply src/rust-crypto/rust-crypto.ts:18]: copy paste error when importing another-json

File: src/rust-crypto/backup.ts [hunk 5]
```diff
@@ -280,6 +308,90 @@ export class RustBackupManager extends TypedEventEmitter<RustBackupCryptoEvents,
...
+     *
+     * If there is an existing backup server side it will be deleted and replaced
+     * by the new one.
+     *
+     * @param signObject - Method that should sign the backup with existing device and
+     * existing identity.
+     * @returns a KeyBackupCreationInfo - All information related to the backup.
+     */
+    public async setupKeyBackup(signObject: (authData: AuthData) => Promise<void>): Promise<KeyBackupCreationInfo> {
+        // Cleanup any existing backup
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply src/rust-crypto/backup.ts:323]: done

File: src/rust-crypto/backup.ts [hunk 6]
```diff
@@ -280,6 +308,90 @@ export class RustBackupManager extends TypedEventEmitter<RustBackupCryptoEvents,
...
+        );
+
+        this.olmMachine.saveBackupDecryptionKey(version.decryptionKey, res.version);
+
+        return {
+            version: res.version,
+            algorithm: version.algorithm,
+            authData: version.auth_data,
+            decryptionKey: version.decryptionKey,
+        };
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply src/rust-crypto/backup.ts:349]: right, updated.

File: spec/integ/crypto/crypto.spec.ts (hunk 1 above)
Comment (BillCarsonFr) [reply spec/integ/crypto/crypto.spec.ts:57]: sry, messed up one of the rebase

File: spec/integ/crypto/crypto.spec.ts (hunk 2 above)
Comment (BillCarsonFr) [reply spec/integ/crypto/crypto.spec.ts]: not sure I touched that, I am a bit confused

**Review by richvdh — APPROVED** [review]
lgtm

**Pull request merged**